capi: support Flatcar Container Linux

dongsupark opened the pull request:
This is a cluster-API image builder for [Flatcar Container Linux](https://www.flatcar-linux.org).

On the Packer side it includes `packer/qemu/qemu-flatcar.json` and `packer/qemu/packer.json`. In `packer/config/` it also includes some other JSON files with configuration for ansible and packages installed with Ansible.

On the Ansible side there's some preliminary playbook in `ansible/`.

## notes

* `images/capi/ansible/roles/setup/tasks/bootstrap-flatcar.yml` installs python under `/opt`.
* `images/capi/ansible/roles/kubernetes/tasks/url.yml` installs artificts of Kubernetes and CNI under `/opt` and `/etc`, mainly because `/usr` is read-only in Flatcar.
* It unpacks only parts of containerd binaries, since Flatcar already has its own built-in containerd.
* To make containerd work correctly, we install Flatcar-specific config files for containerd, so it listens on its containerd socket `/run/docker/libcontainerd/docker-containerd.sock`.
* Since Flatcar has its own containerd socket, we need to also specify the socket name when running commands like `kubeadm config images pull`, `/opt/bin/ctr images import` as well as in `/etc/crictl.yaml`.
* We do not delete `/etc/kubeadm.yml` for Flatcar, because later steps running `kubeadm init` require the config to be in place.

## TODOs

* Make sure it works for other platforms than qemu & AWS

/cc @vbatts @t-lo @dims @detiber @codenrhoden

t-lo commented:
Signed. Can we re-evaluate please?

dongsupark commented:
Rebased, and added AWS AMI support.

CecileRobertMichon commented on images/capi/id_rsa:
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
+MIIEogIBAAKCAQEA6NF8iallvQVp22WDkTkyrtvp9eWW6A8YVr+kz4TjGYe7gHzI
```
@dongsupark was this checked in by mistake?

dongsupark replied:
No, the key is exactly the same key as the [vagrant insecure ssh key](https://github.com/hashicorp/vagrant/blob/master/keys/vagrant). 
It is needed for vagrant tests.

Now I am thinking, it might be better to rename the key to a descriptive name, e.g. `vagrant-insecure-key`.

CecileRobertMichon commented on images/capi/hack/image-build-flatcar.sh:
```diff
@@ -0,0 +1,115 @@
+    make FLATCAR_CHANNEL="$channel" FLATCAR_VERSION="$release" \
+        build-${CAPI_PROVIDER}-flatcar
+    run_vagrant
+elif [[ ${CAPI_PROVIDER} = "aws" ]] || [[ ${CAPI_PROVIDER} = "ami" ]]; then
```
any plans to add the same support for azure?

dongsupark replied:
Sure, we can add Azure support in the future.
Before that, I would like to get some feedback from maintainers about the current PR, to make sure we are doing in the right direction.

dongsupark replied:
Renamed the ssh key file to `vagrant-insecure-key`.

codenrhoden commented on images/capi/ansible/roles/setup/tasks/main.yml:
```diff
@@ -15,6 +15,9 @@
   when: ansible_os_family == "Debian"
 
+- import_tasks: flatcar.yml
+  when: ansible_os_family == "Flatcar"
```
I'm still missing how this fact gets set when `gather_facts` is False. :(

dongsupark replied:
> I'm still missing how this fact gets set when gather_facts is False. :(

`gather_facts` is set to True now.

codenrhoden commented on images/capi/ansible/roles/sysprep/tasks/main.yml:
```diff
@@ -112,13 +121,23 @@
     path: "{{ item.path }}"
   loop: "{{ ssh_host_keys.files }}"
 
+- setup: filter=ansible_env
```
I'm hoping we can sort out the `gather_facts` change so this (and the below change) isn't necessary

dongsupark replied:
> I'm hoping we can sort out the gather_facts change so this (and the below change) isn't necessary

Since `gather_facts` is True now, the filter is also gone.

codenrhoden commented on images/capi/packer/ami/packer.json:
```diff
@@ -1,7 +1,9 @@
     "ansible_common_vars": "",
+    "ansible_distribution": "",
     "ansible_extra_vars": "",
+    "ansible_os_family": "Flatcar",
```
Oh... now I see where this are coming. Yeah, these need to be determined by Ansible, not manually set. :)

codenrhoden replied:
And besides, by setting this here, and then passing it through in `ansible-args.json`, you are effectively setting this for every build in this file. Which  means that Ubuntu, CentOS, and AL2 now have this set to "Flatcar". That won't work.

I desperately need to refactor the AMI builds to be like the OVA builds -- with a single, common builder stanza (with `type: amazon-ebs`), and with each OS separated into it's own .json file for unique vars. The way it's strucutured right now makes changes like this impossiblel to apply to only a single OS. I may work on that this week, because otherwise adding Flatcar in is going to be really difficult since it has so many things that need to be changed that don't need to be changed in any other OS.

dongsupark replied:
> I desperately need to refactor the AMI builds to be like the OVA builds -- with a single, common builder stanza (with type: amazon-ebs), and with each OS separated into it's own .json file for unique vars. The way it's strucutured right now makes changes like this impossiblel to apply to only a single OS. I may work on that this week, because otherwise adding Flatcar in is going to be really difficult since it has so many things that need to be changed that don't need to be changed in any other OS.

All right.
Please let me know when you managed to refactor it.

codenrhoden commented on images/capi/packer/ami/packer.json:
```diff
@@ -209,12 +255,6 @@
-      "type": "goss",
-      "url": "{{user `goss_url`}}",
-      "use_sudo": true,
-      "tests": ["packer/goss/goss.yaml"]
```
Why is this removed? We want to keep this.

dongsupark replied:
> Why is this removed? We want to keep this.

It is still an open issue.

The reason for removing goss for Flatcar is because goss is not able to detect Flatcar as one of the distros at all. So when goss tries install `cloud-init` package, it simply fails.
Have created an [upstream issue](https://github.com/aelsabbahy/goss/issues/596).

On the other hand, it does not seem possible to run goss only for specific distros.
I am not sure what is the path forwards.

dongsupark replied:
As for goss, I ended up creating 2 different `goss.yaml` files, one for Flatcar and the other for non-Flatcar.

Ideally we should be able to make use of templates as [described](https://github.com/aelsabbahy/goss/blob/master/docs/manual.md#examples-2) in goss docs.
However, it is not possible to do so, because [packer-provisioner-goss](https://github.com/YaleUniversity/packer-provisioner-goss) does not take environment variables as input parameters that can be used for such a distinction.

That's why I had to create another `goss.yaml` for Flatcar.

See my WIP work for details: https://github.com/kinvolk/image-builder/pull/4/
Though it could conflict with other WIP PRs like https://github.com/kubernetes-sigs/image-builder/pull/277 .

codenrhoden commented on images/capi/packer/qemu/packer.json:
```diff
@@ -1,12 +1,15 @@
-    "ansible_extra_vars": "ansible_python_interpreter=/usr/bin/python3",
-    "boot_wait": "10s",
+    "ansible_extra_vars": "ansible_python_interpreter=/opt/bin/python",
+    "ansible_os_family": "Flatcar",
```
Same comment as before.

dongsupark replied:
> Same comment as before.

Now `ansible_os_family` should be defined only in Flatcar-specific configs.

codenrhoden commented on images/capi/packer/qemu/packer.json:
```diff
@@ -1,12 +1,15 @@
-    "boot_wait": "10s",
+    "ansible_extra_vars": "ansible_python_interpreter=/opt/bin/python",
+    "ansible_os_family": "Flatcar",
+    "boot_wait": "120s",
```
Just like I said I need to refactor the AMI builds to be structured like OVAs when it comes to multiple OSs, the same applies here. If this really needs to change, it would be better to set it per-OS.

dongsupark replied:
> Just like I said I need to refactor the AMI builds to be structured like OVAs when it comes to multiple OSs, the same applies here. If this really needs to change, it would be better to set it per-OS.

Done. Now `boot_wait` is set to 120s only for Flatcar.

codenrhoden commented on images/capi/packer/qemu/packer.json:
```diff
@@ -1,12 +1,15 @@
     "containerd_version": null,
     "containerd_sha256": null,
     "containerd_url": "https://storage.googleapis.com/cri-containerd-release/cri-containerd-{{user `containerd_version`}}.linux-amd64.tar.gz",
+    "containerd_additional_settings": "",
```
There's no reason to add this here.

dongsupark replied:
> There's no reason to add this here.

Removed.

codenrhoden commented on images/capi/packer/qemu/packer.json:
```diff
@@ -34,7 +37,9 @@
     "kubernetes_container_registry": null,
+    "machine_id_mode": "444",
     "output_directory": "./output/{{user `build_name`}}-kube-{{user `kubernetes_semver`}}",
+    "override_system_hostname": "true",
```
this is unused in Ansible.

dongsupark replied:
> this is unused in Ansible.

Removed.

codenrhoden commented on images/capi/packer/qemu/packer.json:
```diff
@@ -62,13 +67,25 @@
         "{{user `boot_command_prefix`}}",
         "http://{{ .HTTPIP }}:{{ .HTTPPort }}",
-        "{{user `boot_command_suffix`}}"
+        "{{user `boot_command_suffix`}}",
```
As written, This is going to be run for the Ubuntu build as well. I don't think you want that. Instead, this should be added to the `boot_command_suffix` for Flatcar.

dongsupark replied:
> As written, This is going to be run for the Ubuntu build as well. I don't think you want that. Instead, this should be added to the boot_command_suffix for Flatcar.

Done. The boot commands are updated only for Flatcar.

codenrhoden commented on images/capi/packer/qemu/packer.json:
```diff
@@ -88,5 +105,6 @@
     }
-  ]
+  ],
+  "post-processors": ["vagrant"]
```
This is adding it for all OS's as well. I don't think we want that.

dongsupark replied:
> This is adding it for all OS's as well. I don't think we want that.

Done. The vagrant post-processor runs only for Flatcar.

dongsupark commented:
@codenrhoden Thanks a lot for your thorough review.
Will try to address the issues.

dongsupark commented on images/capi/README-flatcar.md:
```diff
@@ -0,0 +1,104 @@
+
+## Overview
+
+This is a cluster-API image builder for [Flatcar Container Linux](https://www.flatcar-linux.org).
```
Done.

dongsupark commented on images/capi/ansible/node.yml:
```diff
@@ -13,6 +13,7 @@
 # limitations under the License.
 ---
 - hosts: all
+  gather_facts: false # not all hosts might be able to run modules yet
```
> I'd much rather see the use of the Shell provisioner in Packer to run before the Ansible provisioner in the case of Flatcar. This could install Python or anything else needed to get the node ready to be provisioned by Ansible. The use of only_on should make this able to happen only for Flatcar as well.

As suggested, I enabled `gather_facts` again. To do that, I used a shell provisioner on the packer side, before running ansible. That way, we can install python without modifying ansible files.

Note, [only_on](https://www.packer.io/docs/provisioners/shell-local#only_on) of a shell-local provisioner is meant to be used for selecting OS, e.g. `linux` or `windows`. We cannot use that for distros like `Flatcar` or `Ubuntu`. Instead, we need to use [only](https://www.packer.io/docs/provisioners/shell#only) of a shell provisioner.

dongsupark commented on images/capi/ansible/roles/containerd/templates/etc/containerd/config-flatcar.toml:
```diff
@@ -0,0 +1,27 @@
+# persistent data location
```
> Rather than have a Flatcar-specific version of this template, I'd rather see the existing template have elements that are added when we are building for Flatcar. As it is, existing customization options for containerd are lost with this change (ability to specify a custom pause image, forcing v2 syntax, ability to append any custom config at the end).

Used jinja2 templates as suggested. It should be now simpler than before.

dongsupark commented on images/capi/ansible/roles/containerd/templates/etc/systemd/system/containerd-flatcar.service:
```diff
@@ -0,0 +1,25 @@
+LimitNOFILE=1048576
+LimitNPROC=infinity
+LimitCORE=infinity
+TasksMax=infinity
```
> I think you mentioned that Flatcar ships with containerd. Is there not already a unit file present for it? I'm concerned about there being confusion or conflicts between this and the additional drop-ins that are put in place today...

Right. Flatcar already has built-in containerd and docker provided by torcx. It basically means that after boots it is not possible to modify any torcx artifacts, including the containerd's systemd unit file as well as the containerd config. However, we need to download containerd binary files, modify its config, so the new containerd points to the built-in gRPC socket provided by Docker. To do that, we need to install a new containerd unit file that points to a new containerd config that has path to the new containerd socket.

I admit it could look ugly, but I was not able to find a better solution.

dongsupark commented on images/capi/ansible/roles/kubernetes/files/usr/lib/systemd/system/kubelet-flatcar.service:
```diff
@@ -0,0 +1,12 @@
+[Unit]
```
> Some comment as before. I'd rather see a unified template than have a second one broken out. That way we can see what is common, and what is unique to each. I'm worried about us wanting to make a common change (say we change RestartSec from 10 to 20) and it only happening in one place instead of two.

Updated it so a single template `kubelet.service` can take a parameter `sysusr_prefix`.

dongsupark commented on images/capi/ansible/roles/kubernetes/files/usr/lib/systemd/system/kubelet.service.d/10-kubeadm-flatcar.conf:
```diff
@@ -0,0 +1,11 @@
+# Note: This dropin only works with kubeadm and kubelet v1.11+
```
> Same.

Updated it so a single template `10-kubeadm.conf` can take a parameter `sysusr_prefix`.

dongsupark commented on images/capi/ansible/roles/kubernetes/tasks/ecrpull.yml:
```diff
@@ -24,4 +24,5 @@
-    state: absent
\ No newline at end of file
+    state: absent
+  when: ansible_os_family != "Flatcar"
```
> Nothing here should be running kubeadm init, that step is run via cloud-init, and put in place by a CAP* controller.

Right.
I just meant, the new script `hack/image-build-flatcar.sh` runs also Vagrant tests including `kubeadm init`.

dongsupark commented on images/capi/packer/qemu/packer.json:
```diff
@@ -1,12 +1,15 @@
     "containerd_sha256": null,
     "containerd_url": "https://storage.googleapis.com/cri-containerd-release/cri-containerd-{{user `containerd_version`}}.linux-amd64.tar.gz",
+    "containerd_additional_settings": "",
+    "containerd_cri_socket": "/run/docker/libcontainerd/docker-containerd.sock",
```
> Instead, maintain the current default valuee of /var/run/containerd/containerd.sock in packer/config/containerd.json, and override it in the singular case that needs it (Flatcar). I would suggest including a flatcar.json var file in the Makefile for that. As it is, these builds will now work for Flatcar but for no other OS's.

Updated it as suggested.

dongsupark commented on images/capi/ansible/roles/kubernetes/tasks/url-flatcar.yml:
```diff
@@ -0,0 +1,129 @@
+# Copyright 2020 The Kubernetes Authors.
```
> This file looks mostly to be a duplicate (Create, download, install, remove...) of url.yml, and I'm not seeing anything that can't be handled with some when: statements. Though I'm admittedly not going line by line. Since the majority of this file is copied, however, and you find that you can't get what you need with simple conditions, I'd still rather see the existing url.yml split into smaller pieces that can then be re-used with import_tasks.

Updated it so a single task `url.yml` can take a parameter `systemd_prefix`.

dongsupark commented on images/capi/ansible/roles/node/tasks/main.yml:
```diff
@@ -12,6 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 ---
+
```
> No need for this whitespace change. We don't put a newline between --- and the first code block in the other files.

Removed the empty line.

dongsupark commented on images/capi/ansible/roles/providers/tasks/aws.yml:
```diff
@@ -18,7 +18,9 @@
-  when: ansible_distribution != "Amazon"
+  when:
+    - ansible_os_family != "Flatcar"
+    - ansible_distribution != "Amazon"
```
> So Flatcar already has this?

No, Flatcar does not have awscli by default.
It is still an open issue.

The reason for doing only for non-Flatcar is because it is not possible to install awscli, which still depends on python 2. At the moment, `bootstrap.sh` installs pypy 3.5 on Flatcar.

There might be two options.
* Make `bootstrap.sh` install also pypy 2.7. Though it did not work when I tried.
* Wait until `awscli` supported python 3.x. See a related [issue](https://github.com/aws-cloudformation/aws-cloudformation-coverage-roadmap/issues/151).

I am not sure what is the path forwards.

dongsupark replied:
Fixed the awscli issue by following the first approach, `Make bootstrap-flatcar.sh install also pypy 2.7`.

To do that, I had to introduce a variable `python_path`, which allows users to specify Flatcar-specific python libs path. Also created python virtualenv `/opt/bin/builder-env`, where the python libraries are installed.

dongsupark commented on images/capi/ansible/roles/setup/defaults/main.yml:
```diff
@@ -12,6 +12,9 @@
 # limitations under the License.
 ---
+# Set the hostname to inventory_hostname
+override_system_hostname: "{{ override_system_hostname }}"
```
> This variable doesn't appear to be used anywhere? A quick Ctrl-F finds it never being used.

Removed the variable `override_system_hostname`.

dongsupark commented on images/capi/ansible/roles/setup/tasks/bootstrap-flatcar.yml:
```diff
@@ -0,0 +1,32 @@
+---
```
> As mentioned before, I feel like bootstrapping Python into Flatcar is better accomplished as a separate provisioner (not with Ansible).

Done

vbatts commented:
will rebase once #289 is merged

dongsupark commented:
Since https://github.com/kubernetes-sigs/image-builder/pull/289 was merged, rebased this PR on top of master.
Fixed bugs around goss and packer 1.6 as well.

dims commented:
/uncc

dongsupark commented:
rebased

moshloop commented on images/capi/Makefile:
```diff
@@ -152,11 +152,20 @@ ABSOLUTE_PACKER_VAR_FILES := $(foreach f,$(abspath $(PACKER_VAR_FILES)),-var-fil
 
+# Set Flatcar Container Linux channel and version if not supplied
+FLATCAR_CHANNEL ?= stable
+FLATCAR_VERSION ?= $(shell hack/image-grok-latest-flatcar-version.sh $(FLATCAR_CHANNEL))
```
I think this should be explicit as otherwise, the output is not deterministic

dongsupark replied:
> I think this should be explicit as otherwise, the output is not deterministic

Good point. Will simply use `stable`.

dongsupark replied:
Done.

moshloop commented on images/capi/ansible/roles/kubernetes/tasks/kubeadmpull.yml:
```diff
@@ -5,9 +5,10 @@
 
 - name: Kubeadm pull images
-  shell: 'kubeadm config images pull --config /etc/kubeadm.yml'
+  shell: 'kubeadm config images pull --config /etc/kubeadm.yml --cri-socket {{ containerd_cri_socket }}'
```
The removal of the `--cri-socket` argument should probably be a separate PR as it will impact all builds.

dongsupark replied:
> The removal of the --cri-socket argument should probably be a separate PR as it will impact all builds.

I supppose you mean `addition of the --cri-socket`.
I can split the cri-socket part into a separate PR if you want.
But then it would be impossible to build Flatcar images at all within this PR, because the default socket `/var/run/containerd/containerd.sock` is not available in Flatcar.
That's why I had to gate the socket name in multiple cases, i.e. `/run/docker/libcontainerd/docker-containerd.sock` for Flatcar and `/var/run/containerd/containerd.sock` for non-Flatcar. Doing that, we can make all distros work.

Does it make sense?

codenrhoden replied:
I know we agreed to merge this PR in as-is, but just to re-iterate the point, I think it would be to create a separate PR that only adds `--cri-socket` and uses the same default. in that PR, you wouldn't be able to build Flatcar because Flatcar would not be added yet. In this PR, you would then expand upon being allowed to pass in different CRI sockets to make it change depending on OS.

moshloop commented on images/capi/hack/Vagrantfile.flatcar:
```diff
@@ -0,0 +1,25 @@
+ENV["TERM"] = "xterm-256color"
```
I am not sure I understand where Vagrant fits in - are these just meant to be helper scripts that are used for testing locally? In which case I am not sure they belong in this repo.

`hack` directories are notoriously difficult to get cleaned up, I would be more comfortable with a link to another example repo.

dongsupark replied:
> I am not sure I understand where Vagrant fits in - are these just meant to be helper scripts that are used for testing locally? In which case I am not sure they belong in this repo.
> hack directories are notoriously difficult to get cleaned up, I would be more comfortable with a link to another example repo.

The `Vagrantfile.flatcar` is meant to be used for testing it locally.
Ok, then I will try to move `Vagrantfile.flatcar` into another place.

dongsupark replied:
Done.

moshloop commented on images/capi/packer/qemu/packer.json:
```diff
@@ -1,7 +1,7 @@
   "variables": {
     "ansible_common_vars": "",
-    "ansible_extra_vars": "ansible_python_interpreter=/usr/bin/python3",
+    "ansible_extra_vars": "",
```
Does the Ubuntu 18.04 QEMU build still work with this?

dongsupark replied:
> Does the Ubuntu 18.04 QEMU build still work with this?

Yes, it works well, no matter whether the `ansible_python_interpreter` is set or not.
However, to avoid confusion, I will bring back the `ansible_python_interpreter=/usr/bin/python3` in `packer/qemu/packer.json`.

dongsupark replied:
Done.

moshloop commented on images/capi/packer/ami/packer.json:
```diff
@@ -13,6 +14,7 @@
     "containerd_version": null,
     "containerd_sha256": null,
     "containerd_url": "https://storage.googleapis.com/cri-containerd-release/cri-containerd-{{user `containerd_version`}}.linux-amd64.tar.gz",
+    "containerd_cri_socket": "/run/docker/libcontainerd/docker-containerd.sock",
```
This doesn't look right - I think the default/standard is `unix:///var/run/containerd/containerd.sock`

dongsupark replied:
> This doesn't look right - I think the default/standard is unix:///var/run/containerd/containerd.sock

Good catch. Will fix it.

dongsupark replied:
Done.

moshloop commented on images/capi/packer/ami/packer.json:
```diff
@@ -37,6 +39,7 @@
     "kubernetes_semver": null,
     "kubernetes_source_type": null,
     "manifest_output": "manifest.json",
+    "python_path": "{{env `PYTHONPATH`}}",
```
Will this not return the PYTHONPATH on the machine running packer, not the machine ansible is SSH'ing into ?

dongsupark replied:
> Will this not return the PYTHONPATH on the machine running packer, not the machine ansible is SSH'ing into ?

You are right. I will set `python_path` simply to `null`.

dongsupark replied:
Done. Set `python_path` to `""`.

moshloop commented on images/capi/ansible/roles/sysprep/tasks/main.yml:
```diff
@@ -119,6 +128,15 @@
   - { path: "/home/{{ ansible_env.SUDO_USER }}/.ssh/authorized_keys" }
+  when: ansible_os_family != "Flatcar"
+
+- name: Remove SSH authorized users for Flatcar
```
I think this may have unintended side-effects as many users will want to customize the image with their SSH keys.

dongsupark replied:
> I think this may have unintended side-effects as many users will want to customize the image with their SSH keys.

I am not sure I understand.

This step is to avoid removing `/home/{{ ansible_env.SUDO_USER }}/.ssh/authorized_keys` in Flatcar, because Flatcar needs to keep `authorized_keys` configured via ignition and update-ssh-keys. If the file is missing, Flatcar machine does not allow any ssh login from normal users.

Are you suggesting removing `authorized_keys` also in Flatcar?
Then I am not sure how we should fix the ssh key issue.

moshloop commented on images/capi/ansible/roles/kubernetes/tasks/main.yml:
```diff
@@ -41,7 +41,7 @@
 
 - import_tasks: kubeadmpull.yml
-  when: kubernetes_source_type == "pkg" and ecr != true
+  when: ecr != true
```
Is this not changing the behavior for all builds?

dongsupark replied:
> Is this not changing the behavior for all builds?

Right. I will update the `when` clause.

moshloop commented on images/capi/ansible/roles/containerd/tasks/main.yml:
```diff
@@ -39,6 +39,45 @@
+    src: /tmp/containerd.tar.gz
+    dest: /opt
+    exclude:
+      - ./usr/local/bin/containerd
```
Is Flatcar running a customized/forked version of containerd? Should we not be overwriting the built-in version with the user-specified version to make the containerd version consistent across builds ?

dongsupark replied:
> Is Flatcar running a customized/forked version of containerd? Should we not be overwriting the built-in version with the user-specified version to make the containerd version consistent across builds ?

Flatcar runs a vanilla containerd, 1.1 for stable and 1.3 for beta/alpha as of now. However, containerd in Flatcar is one of the packages protected by [torcx](https://github.com/flatcar-linux/torcx), a boot-time add-on manager. It means, its built-in binary `/run/torcx/unpack/docker/bin/containerd` is located under a read-only partition `/run/torcx/unpack`, and the mount point is "sealed" through the entire machine's lifetime. So it is not possible overwrite the build-in containerd binary with another binary.

The downloaded binary `/usr/local/bin/containerd` will be kept separately from the torcx-managed binaries.

moshloop replied:
Perhaps I am missing something, but if you are excluding the containerd binaries, then why download the containerd archive at all?  I don't see where `/opt/bin/containerd` is created our used (The systemd unit still uses `${TORCX_BINDIR}/containerd` )

dongsupark replied:
> Perhaps I am missing something, but if you are excluding the containerd binaries, then why download the containerd archive at all? I don't see where /opt/bin/containerd is created our used (The systemd unit still uses ${TORCX_BINDIR}/containerd )


Let me go through the whole process in `ansible/roles/containerd/tasks/main.yml`.

1. download containerd
 : It downloads a tarball from [containerd_url](https://storage.googleapis.com/cri-containerd-release/cri-containerd-1.3.4.linux-amd64.tar.gz), and stores it into `/tmp/containerd.tar.gz`. The tarball includes all the containerd binaries to be unpacked into the target filesystem, including `/opt`, '/etc`, and `/usr`.

2. unpack containerd into /opt
 : It unpacks `/tmp/containerd.tar.gz` into `/opt`, excluding binaries like `/usr/local/bin/containerd` or `/usr/local/sbin/runc`, because they are already provided by Flatcar itself.
 
   On the other hand, it still needs to install other binaries like `/usr/local/bin/containerd-shim` or `/usr/local/bin/crictl`, as they are not included in Flatcar. Obviously it has to install the binaries into the target directory `/opt`, because `/usr` in Flatcar is read-only.

3. unpack containerd into /etc
 : unpacks `/tmp/containerd.tar.gz` into `/etc`, excluding other binaries as well as some unnecessary config files.

We need to go through multiple steps that might look complicated, because we need to install only necessary binaries that are included in the downloaded cri-containerd tarball, without trying to touch anything in read-only partitions.

moshloop replied:
| unpack containerd into /opt

The issue is that we are not unpacking containerd, rather some other bits that are not containerd which breaks an implicit contract: If I provide a specific containerd version to image-builder, I expect that version to be used in the resulting image (I understand *why* you can't use a specific version).

With this patch the installation comprises of a patched containerd binaries supplied by the OS, and some other files supplied by a user-specified containerd version. 

From what I can see in https://github.com/flatcar-linux/coreos-overlay/blob/main/app-emulation/containerd/containerd-9999.ebuild#L61 containerd-shim is already installed?

My suggestions are:

1) If the user-supplied containerd version does not match the containerd version supplied by Flatcar, fail hard
2) Install crictl via https://github.com/kubernetes-sigs/cri-tools
3) Embed and install the containerd config files / units in the playbook

dongsupark replied:
> From what I can see in https://github.com/flatcar-linux/coreos-overlay/blob/main/app-emulation/containerd/containerd-9999.ebuild#L61 containerd-shim is already installed?

Only `containerd-shim` is available via torcx.
Neither `containerd-shim-v1` nor `containerd-shim-v2`.
Without installing both containerd shims into `/opt`, v1 and v2, the image-builder simply does not work.
No matter what, we cannot get rid of unpacking the cri-containerd tarball.

> 1. If the user-supplied containerd version does not match the containerd version supplied by Flatcar, fail hard

It makes sense. I will see what I can do.

> 2. Install crictl via https://github.com/kubernetes-sigs/cri-tools

Do you mean, like downloading a specific tarball like https://github.com/kubernetes-sigs/cri-tools/releases/download/v1.16.1/crictl-v1.16.1-linux-amd64.tar.gz ?
I can do that if you want.
Though I am not sure if that's in the scope of this PR.
In the master branch it already downloads the cri-containerd tarball, which includes its crictl binary.
If you mean creating another PR based on master to make it depend on the upstream cri-tools repo, then I can do that as well.

> 3. Embed and install the containerd config files / units in the playbook

I thought we are already doing that.

moshloop replied:
Can `containerd-shim-v1` and `containerd-shim-v2` not be fixed in upstream emulation-package?

2 & 3 are instead of downloading the containterd tarball, not in addition

moshloop commented on images/capi/packer/qemu/qemu-flatcar.json:
```diff
@@ -0,0 +1,22 @@
+  "ansible_extra_vars": "ansible_python_interpreter=/opt/bin/python",
+  "ansible_os_family": "Flatcar",
+  "boot_wait": "120s",
+  "boot_command_prefix": "curl -sLo .ssh/authorized_keys https://raw.githubusercontent.com/hashicorp/vagrant/master/keys/vagrant.pub<enter>curl -sLo /tmp/ignition.json https://raw.githubusercontent.com/flatcar-linux/flatcar-packer-qemu/master/ignition-vagrant.json<enter>sed -i 's/\"name\": \"vagrant\"/\"name\": \"builder\"/' /tmp/ignition.json<enter>sudo flatcar-install -d /dev/sda -C {{user `cha ...
```
I am very hesitant of this boot command - If you boot an instance with a public IP then anyone will be able to access it, can we not use the SSH key generated on-demand by packer?

dongsupark replied:
> I am very hesitant of this boot command - If you boot an instance with a public IP then anyone will be able to access it, can we not use the SSH key generated on-demand by packer?

I see your point. Those `flatcar-install` commands inside `boot_command_prefix` could look sub-optimal from the first look.

However, we need the commands because it is the only way to specify the ssh public key through the Packer's provisioning steps. AFAIK there is not such a beautiful option for specifying auto-generated ssh keys into the ignition config, as long as we depend on Packer and Ansible. (Terraform seems to have an ignition provider, but that is not an option here.)

You might have noticed that such boot commands are not needed by other cloud providers like AWS, because the ssh keys are configured via other mechanisms. We only need it for local qemu images.

dongsupark commented:
Now this PR depends on Ansible 2.10.

moshloop commented on images/capi/hack/ensure-ansible.sh:
```diff
@@ -20,7 +20,7 @@ set -o pipefail
 source hack/utils.sh
 
-_version="2.8.0"
+_version="2.10.0b1"
```
I think extract this change as well as the other potentially breaking changes into a new PR.  It will make it easier for  others to review and test, once that PR merges it will be easier to review and test just the flatcar changes.

dongsupark replied:
> I think extract this change as well as the other potentially breaking changes into a new PR. It will make it easier for others to review and test, once that PR merges it will be easier to review and test just the flatcar changes.

Ok, then I will try to create another PR to support Ansible 2.10.

figo commented:
/cc

moshloop commented on images/capi/packer/qemu/packer.json:
```diff
@@ -31,14 +34,16 @@
   "builders": [
     {
-      "name": "{{ user `build_name`}}",
+      "name": "flatcar-{{user `channel_name`}}",
```
Can we create a flatcar.json variables file and leave this unchanged? This would override the ubuntu build names

dongsupark replied:
> Can we create a flatcar.json variables file and leave this unchanged? This would override the ubuntu build names.

Good point.
Fixed, thanks.

> Based on the differences between flatcar vs ubuntu it might make sense to make qemu-flatcar.json a packer build file, not only a variable file

I don't have a good idea about how to do it.
It sounds like we might need to change a lot more for that.

moshloop replied:
If you remove the vagrant post-provisioner (which I assume is a development build step) combined with the `only` directive for `"./packer/files/bootstrap-flatcar.sh"` it looks like we can keep them the same

moshloop commented on images/capi/Makefile:
```diff
@@ -159,11 +159,20 @@ ABSOLUTE_PACKER_VAR_FILES := $(foreach f,$(abspath $(PACKER_VAR_FILES)),-var-fil
 
+# Set Flatcar Container Linux channel and version if not supplied
+FLATCAR_CHANNEL ?= stable
+FLATCAR_VERSION ?= $(shell hack/image-grok-latest-flatcar-version.sh stable)
```
This should default to a predefined value so that the builds are deterministic, If users want a non-determistic build that always builds the latest then add a snippet in the docs:

```shell
FLATCAR_VERSION=$(hack/image-grok-latest-flatcar-version.sh stable) make 
```

dongsupark replied:
> FLATCAR_VERSION=$(hack/image-grok-latest-flatcar-version.sh stable) make

Done.

moshloop commented on images/capi/packer/ami/packer.json:
```diff
@@ -113,6 +118,14 @@
+      "environment_vars": [
+        "BUILD_NAME={{user `build_name`}}"
+      ],
+      "script": "./packer/files/bootstrap-flatcar.sh",
```
```suggestion
      "script": "if [ $BUILD_NAME == \"flatcar\" ]; then ./packer/files/bootstrap-flatcar.sh"; fi",
```

dongsupark replied:
>       "script": "if [ $BUILD_NAME == \"flatcar\" ]; then ./packer/files/bootstrap-flatcar.sh"; fi",

Sorry, it is not possible to add arbitrary bash code to `script`.
We need to do in either `script` or `inline`.
Since we need to run a customized bash code with `script` and `execute_command`, the only option is to add the code to `execute_command`.
So I updated the code like that:

```
      "script": "./packer/files/bootstrap-flatcar.sh",
      "execute_command": "BUILD_NAME={{user `build_name`}}; if [[ \"${BUILD_NAME}\" == *\"flatcar\"* ]]; then sudo {{.Vars}} -S -E bash '{{.Path}}'; fi"
```

dongsupark commented:
Updated the PR as @codenrhoden and @moshloop.
It is ready for another review.

Now we simply download the ssh keypair from the official vagrant github repo, to use them for ignition config for Flatcar. Ignition template file `ignition-builder.json` will include a pre-defined string `BUILDERSSHAUTHKEY`, which will be then replaced with the ssh public key given by Packer.

I had also tried using auto-generated ssh keys for Qemu. However I could not, as the Qemu Builder in Packer does not populate a variable like `{{ .SSHPublicKey }}` in templates. Such variables are supported only in Virtualbox or normal cloud providers.
So for Qemu this is the only approach I came up with.

codenrhoden commented on images/capi/packer/ami/packer.json:
```diff
@@ -5,6 +5,7 @@
     "ami_regions": "ap-south-1,eu-west-3,eu-west-2,eu-west-1,ap-northeast-2,ap-northeast-1,sa-east-1,ca-central-1,ap-southeast-1,ap-southeast-2,eu-central-1,us-east-1,us-east-2,us-west-1,us-west-2",
     "ami_users": "",
     "ansible_common_vars": "",
+    "ansible_distribution": "",
```
This is passed through to Ansbile below, but is unused. Remove?

dongsupark replied:
> This is passed through to Ansbile below, but is unused. Remove?

Right. Will remove.

dongsupark replied:
Done. Removed `ansible_distribution`.

codenrhoden commented on images/capi/ansible/node.yml:
```diff
@@ -37,3 +37,4 @@
     HTTP_PROXY: "{{ http_proxy | default('') }}"
     HTTPS_PROXY: "{{ https_proxy | default('') }}"
     NO_PROXY: "{{ no_proxy | default('') }}"
+    PYTHONPATH: "{{ python_path }}"
```
Since this is only being set for Flatcar, can we confirm that we've seen no adverse effects for other builds?

dongsupark replied:
> Since this is only being set for Flatcar, can we confirm that we've seen no adverse effects for other builds?

Yes, I have already tested with other distros like Ubuntu, and I could see it worked well.

codenrhoden commented on images/capi/Makefile:
```diff
@@ -406,6 +429,9 @@ build-node-ova-local-base-rhel-7: ## Builds RHEL 7 Base Node OVA w local hypervi
 
+build-qemu-flatcar-stable: ## Builds Flatcar stable QEMU image
+build-qemu-flatcar-beta: ## Builds Flatcar beta QEMU image
+build-qemu-flatcar-alpha: ## Builds Flatcar alpha QEMU image
```
These three build targets don't exist. The only one that should be documented is `build-qemu-flatcar`.

dongsupark replied:
> These three build targets don't exist. The only one that should be documented is build-qemu-flatcar.

Right. Will keep only `build-qemu-flatcar`.
Will also remove unnecessary variables `FLATCAR_CHANNEL` and `FLATCAR_VERSION` in Makefile.

dongsupark replied:
Only kept `build-qemu-flatcar`.
And kept `FLATCAR_CHANNEL` and `FLATCAR_VERSION` as before, as they are actually needed. Sorry for confusion.

codenrhoden commented on images/capi/ansible/roles/containerd/tasks/main.yml:
```diff
@@ -86,6 +126,16 @@
+- name: stop current containerd service
+  systemd:
+    name: containerd
+    state: stopped
```
Since the task right after this is a restart (even though the name/comment says "start"), why is this step necessary?

dongsupark replied:
> Since the task right after this is a restart (even though the name/comment says "start"), why is this step necessary?

Right. Will remove the stop step.

dongsupark replied:
Removed the `stop` step.

codenrhoden commented on images/capi/ansible/roles/containerd/tasks/main.yml:
```diff
@@ -21,6 +21,11 @@
 - import_tasks: photon.yml
   when: ansible_os_family == "VMware Photon OS"
 
+- name: Fail if containerd version does not match for Flatcar
```
This seems like something that should be taken care of by GOSS. We don't ensure installed versions anywhere else in the Ansible roles. Why do we need this here instead of catching it with GOSS?

dongsupark replied:
> This seems like something that should be taken care of by GOSS. We don't ensure installed versions anywhere else in the Ansible roles. Why do we need this here instead of catching it with GOSS?

I added that part following [suggestion](https://github.com/kubernetes-sigs/image-builder/pull/248#discussion_r485107613) by @moshloop. Moshe wanted to make it fail hard when containerd version does not match.
(Side note: That is the main reason why we should use Flatcar Beta or Alpha, because Flatcar Stable 2605.7.0 still has containerd 1.3. So image-builder fails hard when being used with Stable, because of containerd version mismatch.)

AFAIU image-builder uses goss only for AMI, Azure, OVA. It does not do for Qemu. So I am not sure how we can use goss in that case. 
Note also that goss [does not detect](https://github.com/aelsabbahy/goss/blob/d1e9f590bb65196a46df530b0af3cc3ecd5fd4fa/system/system.go#L184) Flatcar as distro. As a result, using Goss with Flatcar is quite limited, e.g. `service` is possible, but `package` is not possible, obviously because Flatcar does not have a package manager. In the future we could add a full Flatcar support in Goss, but that seemed non-trivial to implement so far.

codenrhoden reviewed (commented):
Hi @dongsupark 

I finally got around to looking at this again. Nice work putting this together! I can definitely tell it was smaller than before. As you've pointed out, we agreed in image-builder office hours to merge this and iterate. I've left a few comments on things I see and would like fixed, but I want to hear your thoughts on what can/should be fixed now versus follow-up PRs.

dongsupark commented:
> I know we agreed to merge this PR in as-is, but just to re-iterate the point, I think it would be to create a separate PR that only adds --cri-socket and uses the same default. in that PR, you wouldn't be able to build Flatcar because Flatcar would not be added yet. In this PR, you would then expand upon being allowed to pass in different CRI sockets to make it change depending on OS.

All right.
Will try to split out the `--cri-socket` part into a separate PR.

dongsupark commented:
Split the cri-socket part into a separate PR https://github.com/kubernetes-sigs/image-builder/pull/433.
Once the new PR got merged, I will rebase this PR.

dongsupark commented:
@codenrhoden Rebased the PR. Please have a look.

codenrhoden commented:
/lgtm

I think this is ready to go! I'll leave this for about 24 hours or so and see if anyone has anything else to say, but I'm happy to merge and iterate. Huge effort @dongsupark!

codenrhoden commented:
/approve

🎉  🚀

dongsupark commented:
/retest

jsturtevant commented:
/ok-to-test
